### PR TITLE
Bug #89350: media type validation

### PR DESCRIPTION
**Pull request by mitchell-fox**
## Description
[PS3.18 (nema.org)](https://dicom.nema.org/medical/dicom/current/output/html/part18.html#sect_8.7):
"An mtp-value can be transmitted either as a token or quoted-string. The quoted and unquoted values are equivalent."

Currently, WADO with `Accept: multipart/related; type=application/dicom` will throw an error. It will accept `Accept: multipart/related; type="application/dicom"`, but they should be equivalent.

##Changes
Altered AcceptContentFilterAttribute.cs to account for media type values with missing quotes such as "multipart/related; type=application/dicom" instead of the usual "multipart/related; type='application/dicom'".

## Addresses
#1396 and [AB#89350](https://microsofthealth.visualstudio.com/f8da5110-49b1-4e9f-9022-2f58b6124ff9/_workitems/edit/89350)

## Testing
Added new tests to AcceptContentFilterAttributeTests.cs.

### CODE REVIEW

File: src/Microsoft.Health.Dicom.Api/Features/Filters/AcceptContentFilterAttribute.cs [hunk 1]
```diff
@@ -83,18 +105,12 @@ public override void OnActionExecuting(ActionExecutingContext context)
                     var stringHeaders = _mediaTypes.Select(x => x.ToString()).ToList();
                     if (split.Any(x => stringHeaders.Contains(x, StringComparer.InvariantCultureIgnoreCase)))
                     {
-                        acceptable = true;
-                        break;
+                        return true;
```
Comment (jovinson-ms) [inline src/Microsoft.Health.Dicom.Api/Features/Filters/AcceptContentFilterAttribute.cs:108]: This whole block can be simplified to: 
```dotnet
if (_mediaTypes.Any(x => x.MatchesMediaType(acceptHeader.MediaType))) 
{
    return true;
}
```

File: src/Microsoft.Health.Dicom.Api.UnitTests/Features/Filters/AcceptContentFilterAttributeTests.cs [hunk 2]
```diff
@@ -130,7 +130,11 @@ public void GivenARequestWithNoAcceptHeader_WhenValidatingTheContentTypeAgainstS
     [InlineData("multipart/related; type=\"application/octet_stream\"", (int)HttpStatusCode.NotAcceptable)]
     [InlineData("multipart/related; type=\"application/octetStream\"", (int)HttpStatusCode.NotAcceptable)]
     [InlineData("multipart/related; type=\"application/dicom\"", null)]
+    [InlineData("multipart/related; type='application/dicom'", (int)HttpStatusCode.NotAcceptable)]
+    [InlineData("multipart/related; type=application/dicom", null)]
     [InlineData("multipart/relATed; type=\"applICation/DICom\"", null)]
+    [InlineData("multipart/relATed; type='applICation/DICom'", (int)HttpStatusCode.NotAcceptable)]
+    [InlineData("multipart/relATed; type=applICation/DICom", null)]
```
Comment (jovinson-ms) [inline src/Microsoft.Health.Dicom.Api.UnitTests/Features/Filters/AcceptContentFilterAttributeTests.cs:137]: We should add a test case with two parameters, like: `multipart/related; type=application/dicom; transfer-syntax="*"`

**Review by jovinson-ms — DISMISSED** [review]

**Comment by jovinson-ms** [comment]
As discussed, the Dicom standard requires the `type` parameter of the `Accept` header to be quoted, since it can include a `/` special character. This is already reflected in our conformance statement.

**Review by shivakumar-ms — DISMISSED** [review]

**Comment by mitchell-fox** [comment]
Committed code for updated PR scope. New code simplifies AcceptContentFilterAttribute so it no longer handles multipart Accept headers as it was never used in the codebase.

**Comment by bcarthic** [comment]
I am surprised none of our E2E tests are broken. Should we add a test?

**Comment by mitchell-fox** [comment]
That's a good idea. I can put them under Microsoft.Health.Dicom.Web.Tests.E2E -> Rest -> AcceptHeaderTests.cs. Should only need a couple E2E tests to validate everything past the updated unit tests.

File: src/Microsoft.Health.Dicom.Api/Features/Filters/AcceptContentFilterAttribute.cs [hunk 3]
```diff
@@ -13,22 +13,17 @@
...
     private const int NotAcceptableResponseCode = (int)HttpStatusCode.NotAcceptable;
-    private const string TypeParameter = "type";
-
-    private readonly bool _allowSingle;
-    private readonly bool _allowMultiple;
 
     private readonly HashSet<MediaTypeHeaderValue> _mediaTypes;
 
-    public AcceptContentFilterAttribute(string[] mediaTypes, bool allowSingle, bool allowMultiple)
+    public AcceptContentFilterAttribute(string[] mediaTypes)
```
Comment (smithago) [inline src/Microsoft.Health.Dicom.Api/Features/Filters/AcceptContentFilterAttribute.cs:26]: Can we confirm this change also works down stream? There are places like dicom-server\src\Microsoft.Health.Dicom.Api\Extensions\MediaTypeHeaderValueExtensions.cs, where the acceptHeader gets converted to internal objects and passed into service, is the type with and without quotes passed in as same value downstream?

Comment (jovinson-ms) [reply src/Microsoft.Health.Dicom.Api/Features/Filters/AcceptContentFilterAttribute.cs:26]: The service currently uses two code paths to handle media type content negotiation - `AcceptContentFilterAttribute` is used only for single part media types like `application/dicom`, and the `GetAcceptHeaders` extension method is only used on controller actions that use the `Produces` attribute for `multipart/related` media types, and then parses to find the media type from the `type` parameter. 

Once we realized that the `AcceptContentFilterAttribute` doesn't handle `multipart/related`, we decided to explicitly handle only single part media types there, and consider a future refactor to handle multipart media type content negotiation at the API layer as well.

File: src/Microsoft.Health.Dicom.Api/Features/Filters/AcceptContentFilterAttribute.cs [hunk 4]
```diff
@@ -46,55 +41,46 @@ public AcceptContentFilterAttribute(string[] mediaTypes, bool allowSingle, bool
...
-                        break;
-                    }
-                }
+    private bool AcceptHeadersContainKnownTypes(HttpRequest request)
+    {
+        IList<MediaTypeHeaderValue> acceptHeaders = ParseAcceptHeaders(request.Headers.Accept);
+
+        foreach (MediaTypeHeaderValue acceptHeader in acceptHeaders)
+        {
+            if (_mediaTypes.Any(x => x.MatchesMediaType(acceptHeader.MediaType)))
```
Comment (jovinson-ms) [inline src/Microsoft.Health.Dicom.Api/Features/Filters/AcceptContentFilterAttribute.cs:66]: As discussed, we will flip this check - the defined media types for this action should be a subset of the Accept header media range, e.g. `application/dicom` is a subset of `application/*` and so will match.

**Review by jovinson-ms — APPROVED** [review]

**Pull request merged**